server/src/test/java/org/elasticsearch/cluster/metadata/MetadataIndexStateServiceBatchingTests.java

```diff
@@ -0,0 +1,232 @@
+import org.elasticsearch.cluster.metadata.IndexMetadata.APIBlock;
+import org.elasticsearch.cluster.metadata.IndexMetadata.State;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.cluster.service.MasterService;
+import org.elasticsearch.common.Priority;
+import org.elasticsearch.core.CheckedRunnable;
+import org.elasticsearch.test.ESSingleNodeTestCase;
+
+import java.util.List;
+import java.util.concurrent.CyclicBarrier;
+import java.util.concurrent.TimeUnit;
+
+import static org.elasticsearch.cluster.metadata.IndexMetadata.INDEX_BLOCKS_WRITE_SETTING;
+import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertAcked;
+import static org.hamcrest.Matchers.containsInAnyOrder;
+import static org.hamcrest.Matchers.hasSize;
+import static org.hamcrest.Matchers.is;
+
+public class MetadataIndexStateServiceBatchingTests extends ESSingleNodeTestCase {
+
+    /*
+     * This class is painfully clever and knows all sorts of implementation details about MetadataIndexStateService. It's hard to get
+     * around that, as we want to make sure that *batching* of opens and closes and adding index blocks will work as expected.
+     *
+     * In the case of opens that's easy enough, we just need to get hands-on enough to make sure that the open tasks are executed in a
+     * single batch -- we do that by blocking up the master service's tasks queue, then running the opens (so they are added to the queue),
+     * and then unblocking the master service again so everything processes in a single batch.
+     *
+     * But closing and adding blocks are both two-step processes where the second step is itself another cluster state update task, and we
+     * want to make sure that those batch up correctly, too. So there we run the same trick as with opens, but twice: we block the
+     * master service in order to batch up the initial tasks, then block it again, so that when the first tasks enqueue the second tasks
+     * they're doing so into a blocked master service, finally the master service is unblocked once more so that the second tasks process
+     * together in a batch.
+     */
+
+    public void testBatchOpenIndices() throws Exception {
+        final var clusterService = getInstanceFromNode(ClusterService.class);
+        final var masterService = clusterService.getMasterService();
+
+        // create some indices
+        createIndex("test-1", client().admin().indices().prepareCreate("test-1"));
+        createIndex("test-2", client().admin().indices().prepareCreate("test-2"));
+        createIndex("test-3", client().admin().indices().prepareCreate("test-3"));
+        ensureGreen("test-1", "test-2", "test-3");
+
+        // and then close them (and wait for the close to actually run)
+        assertAcked(client().admin().indices().prepareClose("test-1", "test-2", "test-3"));
+        assertBusy(() -> assertThat(masterService.pendingTasks(), hasSize(0)));
```

Comment:
Hmm. We apparently need to wait for the `shard-started` tasks to complete, but there might be an intermediate state with zero pending tasks during that process too I think. `ensureGreen` would seem safer? (Also you don't need the previous `ensureGreen`, you can close the index straight away).

Reply:
Sure thing, 69d9f724524cd99876dc7ce73f9fae0fdeedd5c8